### PR TITLE
Prepare for Beta Release

### PR DESCRIPTION
**Pull request by mrnkr**
I did as much as I could to get the project closer to what is proposed in issue #1 - I made the project compile using the latest version of dotnet core, did some integration tests (I'll need some help getting to 100% coverage, I'll admit) and I added an example project.

What's missing from that checklist?

- Travis does not deploy the package tu NuGet, it just runs tests for now.
- Haven't even thought about versioning yet.

I really want this to be published so I hope with this and a bit more effort we'll get there!

Let me know if you see anything I could make better.

### CODE REVIEW

**Comment by alastairtree** [comment]
This is great - have tested the basics and seems to work well. I had to comment out FlushFromCache throwing an exception as that was failing on Patch requests but otherwise the basics of GET POST PATCH and pagination seemed to work really well. What is required to get this merged/completed?

**Comment by bart-degreed** [comment]
Hi @alastairtree thanks for alerting us about this PR, I was unaware of it. It looks promising and I'm willing to provide review feedback in order to meet the same level of quality as the main project.

**Comment by maurei** [comment]
Hi @alastairtree, @mrnkr. Thanks for getting started on this project! A lot has changed in the main library since #1 was opened. First thing that comes to my mind using the [extensively updated integration test suite](https://github.com/json-api-dotnet/JsonApiDotNetCore/tree/master/test/JsonApiDotNetCoreExampleTests/IntegrationTests) in the main project to investigate more thoroughly what needs to be done to get the extension closer to a release. There is still some work to be done. Eg I noticed the mongodb repository is still inheriting from an old `IResourceRepository` interface (for example `FlushFromCache` was removed).

We might want to consider to put this extension in the main library for now, considering that we also don't have a `JsonApiDotNetCore.EntityFrameworkCore` package. But I'm unsure about this.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+[![license][license]](https://github.com/mrnkr/JsonApiDotNetCore.MongoDb/blob/master/LICENSE)
+
+[nuget-image]:https://img.shields.io/nuget/v/JsonApiDotNetCore.MongoDb
+[nuget-url]:https://www.nuget.org/packages/JsonApiDotNetCore.MongoDb
+[downloads-image]:https://img.shields.io/nuget/dt/JsonApiDotNetCore.MongoDb
+[license]:https://img.shields.io/github/license/mrnkr/JsonApiDotNetCore.MongoDb
+
+Plug-n-play implementation of `IResourceRepository<TResource, TId>` allowing you to use MongoDb with your `JsonApiDotNetCore` APIs.
+
+## Installation and Usage
```
Comment (bart-degreed) [inline README.md:16]: I've cloned the repo and tried to run tests. They fail with message:
```
System.TimeoutException
A timeout occured after 30000ms selecting a server using CompositeServerSelector{ Selectors = MongoDB.Driver.MongoClient+AreSessionsSupportedServerSelector, LatencyLimitingServerSelector{ AllowedLatencyRange = 00:00:00.0150000 } }. Client view of cluster state is { ClusterId : "1", ConnectionMode : "Automatic", Type : "Unknown", State : "Disconnected", Servers : [{ ServerId: "{ ClusterId : 1, EndPoint : "Unspecified/localhost:27017" }", EndPoint: "Unspecified/localhost:27017", ReasonChanged: "Heartbeat", State: "Disconnected", ServerVersion: , TopologyVersion: , Type: "Unknown", HeartbeatException: "MongoDB.Driver.MongoConnectionException: An exception occurred while opening a connection to the server.
 ---> System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): No connection could be made because the target machine actively refused it. [::1]:27017
```

This would be the place to describe what needs to be installed. Is there some installer or a docker image that can be used? I'm assuming travisci has mongodb pre-installed?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    }
+}
+```
+
+### Middleware
+
+```cs
+public class Startup
+{
+    public IServiceProvider ConfigureServices(IServiceCollection services) {
```
Comment (bart-degreed) [inline README.md:53]: please put `{` on its own line

File: README.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            options.SerializerSettings.Formatting = Formatting.Indented;
+        }, resources: builder =>
+        {
+            builder.Add<Book, string>();
+        });
+        // ...
+    }
+
+    public void Configure(IApplicationBuilder app)  {
+        app.UseJsonApi();
```
Comment (bart-degreed) [inline README.md:75]: This is insufficient, it needs to be:
```c#
app.UseRouting();
app.UseJsonApi();
app.UseEndpoints(endpoints => endpoints.MapControllers());
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        // ...
+    }
+}
+```
+
+## Limitations
+
+- Relations are not supported (yet)
+- Projections are not supported (yet)
```
Comment (bart-degreed) [inline README.md:84]: What is the problem when using `fields`?

File: src/Example/Properties/launchSettings.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    "iisSettings": {
+      "windowsAuthentication": false,
+      "anonymousAuthentication": true,
+      "iisExpress": {
+        "applicationUrl": "http://localhost:52498",
+        "sslPort": 44343
+      }
+    },
+    "profiles": {
+      "Test": {
```
Comment (bart-degreed) [inline src/Example/Properties/launchSettings.json:12]: Kestrel

File: src/Example/README.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+`dotnet run` to run the project
+
+You can verify the project is running by checking this endpoint:
+`localhost:14141/api/people`
+
+For further documentation and implementation of a JsonApiDotnetCore Application see the documentation or GitHub page:
+
+Repository: https://github.com/json-api-dotnet/JsonApiDotNetCore
+
+Documentation: https://json-api-dotnet.github.io/
```
Comment (bart-degreed) [inline src/Example/README.md:14]: please use: http://www.jsonapi.net

File: src/JsonApiDotNetCore.MongoDb/JsonApiDotNetCore.MongoDb.csproj [hunk 7]
```diff
@@ -1,12 +1,29 @@
...
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <VersionPrefix>4.0.0-beta1</VersionPrefix>
+    <TargetFramework>$(NetCoreAppVersion)</TargetFramework>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <PackageTags>jsonapi;json:api;dotnet;core;mongodb</PackageTags>
+    <Description>Persistence layer implementation for use of mongodb in applications using JsonApiDotNetCore</Description>
```
Comment (bart-degreed) [inline src/JsonApiDotNetCore.MongoDb/JsonApiDotNetCore.MongoDb.csproj:11]: 'MongoDB' seems to be the official casing, according to https://www.mongodb.com/

File: src/JsonApiDotNetCore.MongoDb/JsonApiDotNetCore.MongoDb.csproj [hunk 8]
```diff
@@ -1,12 +1,29 @@
...
+    <PackageProjectUrl>https://github.com/mrnkr/JsonApiDotNetCore.MongoDb</PackageProjectUrl>
```
Comment (bart-degreed) [inline src/JsonApiDotNetCore.MongoDb/JsonApiDotNetCore.MongoDb.csproj:12]: I think this should eventually be changed to json-api-dotnet

File: src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+{
+    public class MongoEntityRepository<TResource, TId>
+        : IResourceRepository<TResource, TId>
+        where TResource : class, IIdentifiable<TId>
+    {
+        private readonly IMongoDatabase _db;
+        private readonly ITargetedFields _targetedFields;
+        private readonly IResourceGraph _resourceGraph;
+        private readonly IResourceFactory _resourceFactory;
+        private readonly IEnumerable<IQueryConstraintProvider> _constraintProviders;
```
Comment (bart-degreed) [inline src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:26]: This field is unused and can be removed.

File: src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace JsonApiDotNetCore.MongoDb
+{
+    public class MongoEntityRepository<TResource, TId>
+        : IResourceRepository<TResource, TId>
+        where TResource : class, IIdentifiable<TId>
+    {
+        private readonly IMongoDatabase _db;
+        private readonly ITargetedFields _targetedFields;
+        private readonly IResourceGraph _resourceGraph;
```
Comment (bart-degreed) [inline src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:24]: This can be `IResourceContextProvider` instead.

File: src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            {
+                Filter = topFilter
+            };
+
+            var query = ApplyQueryLayer(layer);
+            return query.CountAsync();
+        }
+
+        public virtual Task CreateAsync(TResource resource) =>
+            Collection.InsertOneAsync(resource);
```
Comment (bart-degreed) [inline src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:58]: Once you've updated to the latest bits of JADNC, be sure to pass along `CancellationToken`s here.

File: src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+        public virtual async Task<bool> DeleteAsync(TId id)
+        {
+            var result = await Collection.DeleteOneAsync(Builders<TResource>.Filter.Eq(e => e.Id, id));
+            return result.IsAcknowledged && result.DeletedCount > 0;
```
Comment (bart-degreed) [inline src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:63]: After updating to latest bits, throw a `DataStoreUpdateException` if delete fails.

File: src/JsonApiDotNetCore.MongoDb/Queries/Internal/QueryableBuilding/QueryableBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+using JsonApiDotNetCore.Queries.Expressions;
+using JsonApiDotNetCore.Queries.Internal.QueryableBuilding;
+using JsonApiDotNetCore.Resources;
+
+namespace JsonApiDotNetCore.MongoDb.Queries.Internal.QueryableBuilding
+{
+    /// <summary>
+    /// Drives conversion from <see cref="QueryLayer"/> into system <see cref="Expression"/> trees.
+    /// </summary>
+    public sealed class QueryableBuilder
```
Comment (bart-degreed) [inline src/JsonApiDotNetCore.MongoDb/Queries/Internal/QueryableBuilding/QueryableBuilder.cs:14]: Instead of copy/pasting this class, you can set `layer.Projection = null` before calling the built-in version to achieve the same result.

File: test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTestContext.cs [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+            var responseText = await responseMessage.Content.ReadAsStringAsync();
+            var responseDocument = DeserializeResponse<TResponseDocument>(responseText);
+
+            return (responseMessage, responseDocument);
+        }
+
+        private string SerializeRequest(object requestBody) => 
+            requestBody is string stringRequestBody
+                ? stringRequestBody
+                : JsonConvert.SerializeObject(requestBody);
```
Comment (bart-degreed) [inline test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTestContext.cs:95]: We recently fixed a bug that would send request body "null" on GET requests, which I think you should take (see https://github.com/json-api-dotnet/JsonApiDotNetCore/pull/897).

File: test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/CreatingResourcesTests.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        public Task InitializeAsync() => Task.CompletedTask;
+
+        public Task DisposeAsync() => _testContext.RunOnDatabaseAsync(db => db.DropCollectionAsync(nameof(Book)));
+
+        [Fact]
+        public async Task ShouldCreateResource()
+        {
+            var route = "/api/Books";
+            var book = _bookFaker.Generate();
+            var resource = new
```
Comment (bart-degreed) [inline test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/CreatingResourcesTests.cs:48]: This is not a resource, it is a request body. So I think this var should be renamed to `requestBody` instead.

File: test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/CreatingResourcesTests.cs [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+                        category = book.Category,
+                        author = book.Author
+                    }   
+                }
+            };
+            
+            var (httpResponse, responseDocument) = await _testContext.ExecutePostAsync<Document>(route, resource);
+            _createdBookId = responseDocument.Data is ResourceObject resourceObject ? resourceObject.Id : null;
+                
+            Assert.Equal(HttpStatusCode.Created, httpResponse.StatusCode);
```
Comment (bart-degreed) [inline test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/CreatingResourcesTests.cs:66]: We're in the process of adopting [FluentAssertions](https://fluentassertions.com/), would be great if you can use that too. It produces clearer failure messages and abstracts from xUnit syntax.

To check for status code, we have a custom extension that dumps the full response on mismatch.

File: test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/DeletingResourcesTests.cs [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+using Example;
+using Example.Models;
+using JsonApiDotNetCore.Serialization.Objects;
+using Microsoft.Extensions.DependencyInjection;
+using MongoDB.Driver;
+using Xunit;
+
+namespace JsonApiDotNetCore.MongoDb.Example.Tests.IntegrationTests
+{
+    public sealed class DeletingResourcesTests : IClassFixture<IntegrationTestContext<Startup>>, IAsyncLifetime
```
Comment (bart-degreed) [inline test/JsonApiDotNetCore.MongoDb.Example.Tests/IntegrationTests/DeletingResourcesTests.cs:13]: We've recently done major work on improving the tests for create/update/delete, so we now have over 200 integration tests for that.

I think you'd benefit from taking a subset of them. A great amount can be left out, because they check on invalid request bodies. And there are lots of relationship tests.

But others, like delete-not-found, create-already-exists etc. are things you need to handle properly for MongoDb, so this repo would benefit from having those tests.

**Review by bart-degreed — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution! We're going to release JADNC v4-final soon, so it would be great if you can update to that. I've added several inline comments, would be great if you can take a look. Feel free to speak up if you disagree with anything.

**Comment by mrnkr** [comment]
@bart-degreed @maurei Thank you for all your feedback! I'll go through all the comments as soon as I can. @alastairtree thanks for bringing this PR back to life, I'd forgotten about it myself 😂

File: README.md (hunk 4 above)
Comment (mrnkr) [reply README.md:84]: The problem I ran into was that since I was basing myself off of the original `EntityFrameworkCoreRepository` and its dependencies, I found that your `QueryableBuilder` was using the `SelectClauseBuilder` class which depends on some classes exposed by EFCore so I could not bring it into the solution.

I know there is another way I just haven't taken the time to figure it out yet and suggestions are more than welcome.

File: src/JsonApiDotNetCore.MongoDb/Queries/Internal/QueryableBuilding/QueryableBuilder.cs (hunk 13 above)
Comment (mrnkr) [reply src/JsonApiDotNetCore.MongoDb/Queries/Internal/QueryableBuilding/QueryableBuilder.cs:14]: I tried this, it sounds like the right thing to do, but I run into the same problem I had when I was trying to use projections... In this case I can narrow it down to one parameter that's required by `QueryableBuilder`'s constructor which is `IModel model` with `IModel` being exposed by EFCore.

What can we do about this??

File: JsonApiDotNetCore.MongoDb.sln [hunk 18]
```diff
@@ -1,44 +1,86 @@
...
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Example", "src\Example\Example.csproj", "{600A3E66-E63F-427D-A991-4CD2067041F9}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "JsonApiDotNetCore.MongoDb", "src\JsonApiDotNetCore.MongoDb\JsonApiDotNetCore.MongoDb.csproj", "{FD312677-2A62-4B8F-A965-879B059F1755}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{19A533AA-E006-496D-A476-364DF2B637A1}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "JsonApiDotNetCore.MongoDb.IntegrationTests", "test\JsonApiDotNetCore.MongoDb.IntegrationTests\JsonApiDotNetCore.MongoDb.IntegrationTests.csproj", "{5C59FDFE-F079-4015-9975-FEFA766F0787}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "JsonApiDotNetCore.MongoDb.Example.Tests", "test\JsonApiDotNetCore.MongoDb.Example.Tests\JsonApiDotNetCore.MongoDb.Example.Tests.csproj", "{24CE53FA-9C49-4E20-A060-4A43DFB8C8F1}"
+EndProject
```
[the remark this answers is not in the document]
Comment (mrnkr) [reply JsonApiDotNetCore.MongoDb.sln:17]: Actually the one I called `JsonApiDotNetCore.MongoDb.IntegrationTests` does connect to a MongoDb instance to run so it qualifies as integration tests, right? At least that was the reasoning behind my naming the project that way.

**Comment by mrnkr** [comment]
I started working on the comments just now, most of them I've already taken care of but I broke all my tests so I'm obviously fixing that before I merge to master and update this PR.

File: README.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+[license]:https://img.shields.io/github/license/mrnkr/JsonApiDotNetCore.MongoDb
+
+Plug-n-play implementation of `IResourceRepository<TResource, TId>` allowing you to use MongoDb with your `JsonApiDotNetCore` APIs.
+
+## Installation and Usage
+
+### Models
+
+```cs
+public sealed class Book : IIdentifiable<string>
```
[the remark this answers is not in the document]
Comment (mrnkr) [reply README.md:21]: I tried but it caused an error, let me show you what I did:

```cs
public sealed class Book : Identifiable<string>
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    [Attr]
    public new string Id { get; set; }

    // ... The other props ...
    
    [BsonIgnore]
    public new string StringId { get => Id; set => Id = value; }
}
```

The error this caused is `MongoDB.Bson.BsonSerializationException: The property 'Id' of type 'Example.Models.Book' cannot use element name '_id' because it is already being used by property 'Id' of type 'JsonApiDotNetCore.Resources.Identifiable`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]'.`.

Overriding the property this way seems to be causing problems for the MongoDb driver to map the fields properly, using the interface, however, this does not happen so I take it as a win but let me know your thoughts on this.

Just for the record, here is the complete error log:

```
fail: JsonApiDotNetCore.Middleware.ExceptionHandler[0]
      The property 'Id' of type 'Example.Models.Book' cannot use element name '_id' because it is already being used by property 'Id' of type 'JsonApiDotNetCore.Resources.Identifiable`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]'.
MongoDB.Bson.BsonSerializationException: The property 'Id' of type 'Example.Models.Book' cannot use element name '_id' because it is already being used by property 'Id' of type 'JsonApiDotNetCore.Resources.Identifiable`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]'.
   at BsonClassMap MongoDB.Bson.Serialization.BsonClassMap.Freeze()
   at BsonClassMap MongoDB.Bson.Serialization.BsonClassMap.LookupClassMap(Type classType)
   at IBsonSerializer MongoDB.Bson.Serialization.BsonClassMapSerializationProvider.GetSerializer(Type type, IBsonSerializerRegistry serializerRegistry)
   at IBsonSerializer MongoDB.Bson.Serialization.BsonSerializerRegistry.CreateSerializer(Type type)
   at TValue System.Collections.Concurrent.ConcurrentDictionary<TKey, TValue>.GetOrAdd(TKey key, Func<TKey, TValue> valueFactory)
   at IBsonSerializer MongoDB.Bson.Serialization.BsonSerializerRegistry.GetSerializer(Type type)
   at IBsonSerializer<T> MongoDB.Bson.Serialization.BsonSerializerRegistry.GetSerializer<T>()
   at new MongoDB.Driver.MongoCollectionImpl<TDocument>(IMongoDatabase database, CollectionNamespace collectionNamespace, MongoCollectionSettings settings, ICluster cluster, IOperationExecutor operationExecutor)
   at IMongoCollection<TDocument> MongoDB.Driver.MongoDatabaseImpl.GetCollection<TDocument>(string name, MongoCollectionSettings settings)
   at IMongoCollection<TResource> JsonApiDotNetCore.MongoDb.MongoEntityRepository<TResource, TId>.get_Collection() in /Users/mrnkr/Documents/code/JsonApiDotNetCore.MongoDb/src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:line 39
   at IMongoQueryable<TResource> JsonApiDotNetCore.MongoDb.MongoEntityRepository<TResource, TId>.get_Entities() in /Users/mrnkr/Documents/code/JsonApiDotNetCore.MongoDb/src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:line 40
   at IMongoQueryable<TResource> JsonApiDotNetCore.MongoDb.MongoEntityRepository<TResource, TId>.ApplyQueryLayer(QueryLayer layer) in /Users/mrnkr/Documents/code/JsonApiDotNetCore.MongoDb/src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:line 117
   at Task<int> JsonApiDotNetCore.MongoDb.MongoEntityRepository<TResource, TId>.CountAsync(FilterExpression topFilter, CancellationToken cancellationToken) in /Users/mrnkr/Documents/code/JsonApiDotNetCore.MongoDb/src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:line 53
   at TRet System.Dynamic.UpdateDelegates.UpdateAndExecute3<T0, T1, T2, TRet>(CallSite site, T0 arg0, T1 arg1, T2 arg2)
   at async Task<int> JsonApiDotNetCore.Repositories.ResourceRepositoryAccessor.CountAsync<TResource>(FilterExpression topFilter, CancellationToken cancellationToken) in C:/projects/jsonapidotnetcore/src/JsonApiDotNetCore/Repositories/ResourceRepositoryAccessor.cs:line 47
   at async Task<IReadOnlyCollection<TResource>> JsonApiDotNetCore.Services.JsonApiResourceService<TResource, TId>.GetAsync(CancellationToken cancellationToken) in C:/projects/jsonapidotnetcore/src/JsonApiDotNetCore/Services/JsonApiResourceService.cs:line 67
   at async Task<IActionResult> JsonApiDotNetCore.Controllers.BaseJsonApiController<TResource, TId>.GetAsync(CancellationToken cancellationToken) in C:/projects/jsonapidotnetcore/src/JsonApiDotNetCore/Controllers/BaseJsonApiController.cs:line 100
   at async Task<IActionResult> JsonApiDotNetCore.Controllers.JsonApiController<TResource, TId>.GetAsync(CancellationToken cancellationToken) in C:/projects/jsonapidotnetcore/src/JsonApiDotNetCore/Controllers/JsonApiController.cs:line 51
   at async ValueTask<IActionResult> Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor+TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
   at TResult System.Runtime.CompilerServices.ValueTaskAwaiter<TResult>.GetResult()
   at async Task Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()+Awaited(?)
   at async Task Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()+Awaited(?)
   at void Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Task Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
   at Task Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
   at async Task Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeNextExceptionFilterAsync()+Awaited(?)
```

File: README.md (hunk 1 above)
Comment (mrnkr) [reply README.md:16]: I've only just fixed the tests after migrating to the latest version of JADNC and I can confirm all that's necessary is to have a running instance of mongodb on `localhost:27017` with no authentication required. I am personally rocking a docker image for this - `mongo:latest`.

I'm planning on doing as @alastairtree suggested anyway and use mongo2go on my tests so that this is not necessary anymore.

File: README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (bart-degreed) [reply README.md:21]: `new` breaks inheritance, as if the property exists twice. Did you try 'override' instead?

File: README.md (hunk 1 above)
Comment (bart-degreed) [reply README.md:16]: The point is that I have zero experience with mongodb, so 'just have a runnning instance' does not tell me what to do to make that happen. A docker cmdline or choco install command or something similar would help (see [here](https://github.com/json-api-dotnet/JsonApiDotNetCore#testing) for example).

File: README.md (hunk 4 above)
Comment (bart-degreed) [reply README.md:84]: > ...so I could not bring it into the solution.

I don't think that is correct. JsonApiDotNetCore has a dependency on EF Core, and since JsonApiDotNetCore.MongoDb depends on JsonApiDotNetCore, it indirectly also depends on EF Core. NuGet dependencies are transitive, so this should not prevent you from accessing EF Core types. We may split things up in the future, but have no concrete plans for that at the moment.

Easiest solution for now would be to just create a dummy model (you cannot pass `null`), then ensure that `layer.Projection` is always `null` so the dummy model is never used.

```c#
private sealed class DummyModel : IModel
{
    public static IModel Instance { get; } = new DummyModel();

    private DummyModel()
    {
    }

    public IAnnotation FindAnnotation(string name)
    {
        throw new NotImplementedException();
    }

    public IEnumerable<IAnnotation> GetAnnotations()
    {
        throw new NotImplementedException();
    }

    public object this[string name] => throw new NotImplementedException();

    public IEnumerable<IEntityType> GetEntityTypes()
    {
        throw new NotImplementedException();
    }

    public IEntityType FindEntityType(string name)
    {
        throw new NotImplementedException();
    }

    public IEntityType FindEntityType(string name, string definingNavigationName, IEntityType definingEntityType)
    {
        throw new NotImplementedException();
    }
}

// ...

var builder = new QueryableBuilder(source.Expression, source.ElementType, typeof(Queryable), nameFactory, _resourceFactory, _resourceGraph, DummyModel.Instance);
```

This should still give you proper sparse fieldsets in responses, in the sense that only the subset of requested fields is sent back to the client. It just does not optimize the SQL to only fetch the subset of fields. I don't know if that's even possible using MongoDB.

Note that `IncludeBuilder` also depends on an EF Core extension method. If that something MongoDb just ignores or haven't you tried that yet?

File: JsonApiDotNetCore.MongoDb.sln (hunk 18 above)
[the remark this answers is not in the document]
Comment (bart-degreed) [reply JsonApiDotNetCore.MongoDb.sln:17]: [Integration testing](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-5.0) spins up a webserver, sends a HTTP request to that and asserts on the HTTP response. To make that work efficiently, a light-weight or in-memory webserver/database is often used.

A typical integration test in JsonApiDotNetCore looks like the following:
1. setup infrastructure (base class or fixture)
2. seed database with predefined data
3. execute request
4. assert on response (match with what we put in the database upfront)
5. optional: if changes are expected, assert on changed rows in database

Aside to that, we have (a limited set of) unit tests. They test a single class (or a group of closely related classes) in isolation. For example, parsing a query string into expression objects. This is useful when there are many combinations to try (faster). The downside is it may be testing logic that is never used that way in the request execution pipeline, giving a false sense of quality.

File: README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (mrnkr) [reply README.md:21]: I hadn't tried it, no, but I ran into the same error when I did :(
Also, this added another issue, I personally do not want to have MongoDb persist the `StringId` field but I cannot `override` it in order to add the `[BsonIgnore]` annotation because it ain't `virtual`.

**Comment by mrnkr** [comment]
Hello everyone, just wanted to leave you guys an update on this.

I have taken 175 tests from the JADNC repo and 168 of them are passing already. The reimplementation of the repo with the new version of the library is ready too, no relations support yet but all the rest is done.

Basically what I have left to figure out is support for filtering with variables, like `/filterableResources?filter=equals(someNullableInt32,someInt32)`, resource metadata is broken in my implementation and I'm not sure why and, last but not least, I'm failing one test for sparse fieldsets because my implementation of `IModel` is just a dummy and the test I'm failing requires it not to be (test is `Retrieves_all_properties_when_fieldset_contains_readonly_attribute`).

All this is in my fork, in the branch `fix/review_fixes` and I'll get it merged as soon as I get these last few things right.

**Comment by alastairtree** [comment]
Thank you for the update! Sounds really promising. Happy holidays.

**Comment by mrnkr** [comment]
Hello everyone!

Since the PR was getting too big I decided to push off the changes I mentioned before for the time being, this is going to be a nightmare to review as it stands...

Basically this is working fine with the exceptions of:
* filters with variables like `/filterableResources?filter=equals(someNullableInt32,someInt32)` which MongoDb.Driver isn't able to translate for some reason
* resource metadata is just not being used
* `Retrieves_all_properties_when_fieldset_contains_readonly_attribute` needs to be fixed
* relations

Hopefully we can still get this merged and I'll work on these fixes in future PRs, hopefully making it easier to review each of them individually!

Let me know if there is anything that needs fixing!

**Comment by bart-degreed** [comment]
Hi @mrnkr, you've obviously put a lot of time and effort into all of this, thanks for that!

I've skimmed through the changes in your master branch. What I noticed is that lots of tests are commented out because relationships are not supported. Do you have any thoughts on whether it is feasible to add support for that?

I'm happy to see you're working on this and I have been thinking about how to move forward with getting your changes integrated. I believe every commit should improve the codebase, so I'd prefer to not merge lots of commented-out code that's going to be fixed at a later time. We all have lives and shifting priorities, so the 'later' may never come, leaving the project in a somewhat broken state.

Therefore I'd like to propose you submit the following PRs, which I can review and take, one at a time:

1. Update to the latest version of JADNC with .NET Core 3.1. This would include your MongoDB repository and basic getting-started instructions, but without the example/getting-started/test projects. Sort of like: we just updated the codebase to make it run, no guarantees about what actually works.

2. Add checks to the MongoDB repository that fail with an error whenever relationships are used. And add the example and test projects. But instead of commented-out tests, this would contain tests that assert an appropriate error is returned when using relationships. From what I read online, it looks like MongoDB requires IDs to always be strings. So we'll also need a test covering that. And there may be other corner cases to catch, such as the nullable type conversion. This PR would pin the current behavior, identifying what works and what does not.

3. Add some form of relationships support, depending on what is possible. This would replace the tests from PR 2 that produce an error with actual working code.

Something else: I think it would be easier for you to work in separate branches on your fork and keep your master branch in sync with upstream. That way, you can easily merge changes from master into your feature branches.

**Comment by mrnkr** [comment]
@bart-degreed I agree with everything you said, I'll close this PR and do those other smaller ones.

In terms of relationships, I'm pretty sure it is possible but yesterday I did try looking into it and found that it was very hard, at least for me. Will not be trivial and maybe it won't be a very performant implementation since I don't see a way to include referenced documents in a query in one call, it would be a O(n) operation, and an async one at that, so it does not look like something I'd want to use in production...

I agree about the commented code issue, I did think of those as a reminder to tackle them later but it is true later may never come even though I hope it will.
